File: android/sdk/src/main/java/com/viant/forgeandroid/ui/ScheduleEditorRenderer.kt
package com.viant.forgeandroid.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.viant.forgeandroid.runtime.ContainerDef
import com.viant.forgeandroid.runtime.ForgeRuntime
import com.viant.forgeandroid.runtime.ScheduleEditorSpec
import com.viant.forgeandroid.runtime.WindowContext
import com.viant.forgeandroid.runtime.WorkflowPrimitiveRuntime
import java.time.Instant
import java.time.ZoneId

@Composable
internal fun ScheduleEditorRenderer(runtime: ForgeRuntime, window: WindowContext, container: ContainerDef) {
    val spec = container.scheduleEditor ?: return
    val dataSourceRef = spec.dataSourceRef ?: container.dataSourceRef ?: return
    val context = window.contextOrNull(dataSourceRef) ?: return
    val sourceRows by context.collection.flow.collectAsState(initial = context.collection.peek())
    var rows by remember(dataSourceRef) { mutableStateOf(sourceRows.map { it.toMutableMap() }) }
    var dirty by remember(dataSourceRef) { mutableStateOf(false) }
    val validation = WorkflowPrimitiveRuntime.validateSchedule(rows, spec)

    LaunchedEffect(dataSourceRef) { if (context.dataSource.autoFetch != false) context.fetchCollection() }
    LaunchedEffect(sourceRows) { if (!dirty) rows = sourceRows.map { it.toMutableMap() } }

    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        rows.forEachIndexed { index, row ->
            Column(
                verticalArrangement = Arrangement.spacedBy(8.dp),
                modifier = Modifier.fillMaxWidth().background(MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.4f), RoundedCornerShape(10.dp)).padding(10.dp)
            ) {
                Text("Schedule ${index + 1}", style = MaterialTheme.typography.titleMedium)
                ScheduleField("Start (ISO-8601)", row[spec.startField ?: "start"]?.toString().orEmpty()) { value ->
                    rows = updateScheduleRow(rows, index, spec.startField ?: "start", value); dirty = true
                }
                ScheduleField("End (ISO-8601)", row[spec.endField ?: "end"]?.toString().orEmpty()) { value ->
                    rows = updateScheduleRow(rows, index, spec.endField ?: "end", value); dirty = true
                }
                spec.timeZoneField?.let { field ->
                    ScheduleField("Time zone", row[field]?.toString() ?: ZoneId.systemDefault().id) { value ->
                        rows = updateScheduleRow(rows, index, field, value); dirty = true
                    }
                }
                validation.errors.filter { it.index == index }.forEach { error ->
                    Text(scheduleErrorMessage(error.code), color = MaterialTheme.colorScheme.error, style = MaterialTheme.typography.bodySmall)
                }
                if (spec.allowRemove != false) {
                    Button(onClick = { rows = rows.filterIndexed { rowIndex, _ -> rowIndex != index }.map { it.toMutableMap() }; dirty = true }) { Text("Remove") }
                }
            }
        }
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            if (spec.allowAdd != false) Button(onClick = { rows = rows + newScheduleRow(spec); dirty = true }) { Text("Add schedule") }
            spec.mutation?.let { command ->
                MutationCommandButton(
                    runtime, window, context, command,
                    labelOverride = command.label ?: "Save schedule",
                    extras = mapOf("rows" to rows),
                    externallyDisabled = !validation.valid || !dirty
                )
            }
        }
    }
}

@Composable
private fun ScheduleField(label: String, value: String, onChange: (String) -> Unit) {
    OutlinedTextField(value = value, onValueChange = onChange, label = { Text(label) }, singleLine = true, modifier = Modifier.fillMaxWidth())
}

private fun updateScheduleRow(rows: List<Map<String, Any?>>, index: Int, field: String, value: Any?): List<MutableMap<String, Any?>> =
    rows.mapIndexed { rowIndex, row -> row.toMutableMap().apply { if (rowIndex == index) this[field] = value } }

private fun newScheduleRow(spec: ScheduleEditorSpec): MutableMap<String, Any?> {
    val now = Instant.now()
    return mutableMapOf<String, Any?>(
        (spec.startField ?: "start") to now.toString(),
        (spec.endField ?: "end") to now.plusSeconds(3_600).toString()
    ).apply { spec.timeZoneField?.let { this[it] = ZoneId.systemDefault().id } }
}

private fun scheduleErrorMessage(code: String): String = when (code) {
    "invalid_date" -> "Enter a valid start and end time."
    "invalid_range" -> "End time must be after start time."
    "min_duration" -> "The schedule is shorter than the minimum duration."
    "overlap" -> "This schedule overlaps another entry."
    "timezone" -> "Resolve the time-zone or daylight-saving-time value."
    else -> "The schedule is invalid."
}

File: ios/Sources/ForgeIOSUI/ScheduleEditorRenderer.swift
import SwiftUI
import ForgeIOSRuntime

struct ScheduleEditorRenderer: View {
    let runtime: ForgeRuntime
    let window: WindowContext
    let container: ContainerDef

    @State private var rows: [[String: JSONValue]] = []
    @State private var dirty = false

    private var spec: ScheduleEditorSpec { container.scheduleEditor! }
    private var dataSourceRef: String { spec.dataSourceRef ?? container.dataSourceRef ?? "" }
    private var validation: ScheduleValidationResult { WorkflowPrimitiveRuntime.validateSchedule(rows: rows, spec: spec) }

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            ForEach(rows.indices, id: \.self) { index in
                VStack(alignment: .leading, spacing: 8) {
                    HStack {
                        Text("Schedule \(index + 1)").font(.headline)
                        Spacer()
                        if spec.allowRemove != false {
                            Button(role: .destructive) { rows.remove(at: index); dirty = true } label: { Image(systemName: "trash") }
                                .accessibilityLabel("Remove schedule \(index + 1)")
                        }
                    }
                    DatePicker("Start", selection: dateBinding(index: index, field: spec.startField ?? "start"), displayedComponents: [.date, .hourAndMinute])
                    DatePicker("End", selection: dateBinding(index: index, field: spec.endField ?? "end"), displayedComponents: [.date, .hourAndMinute])
                    if let zoneField = spec.timeZoneField {
                        TextField("Time zone", text: stringBinding(index: index, field: zoneField, fallback: TimeZone.current.identifier))
                            .textFieldStyle(.roundedBorder)
                    }
                    ForEach(validation.errors.filter { $0.index == index }, id: \.code) { error in
                        Text(scheduleErrorMessage(error.code)).font(.caption).foregroundStyle(.red)
                    }
                }
                .padding(10)
                .background(.secondary.opacity(0.06), in: RoundedRectangle(cornerRadius: 10))
                .environment(\.timeZone, rowTimeZone(index: index))
            }
            HStack(spacing: 8) {
                if spec.allowAdd != false {
                    Button("Add schedule", systemImage: "plus") { addRow() }
                }
                if let command = spec.mutation {
                    MutationCommandButton(
                        runtime: runtime,
                        window: window,
                        sourceDataSourceRef: dataSourceRef,
                        command: command,
                        labelOverride: command.label ?? "Save schedule",
                        extras: ["rows": .array(rows.map(JSONValue.object))],
                        externallyDisabled: !validation.valid || !dirty
                    )
                }
            }
        }
        .task(id: "\(window.windowID)#\(dataSourceRef)") { await observeRows() }
        .accessibilityIdentifier("forge-schedule-editor")
    }

    @MainActor
    private func observeRows() async {
        rows = await runtime.dataSourceCollection(windowID: window.windowID, dataSourceRef: dataSourceRef)
        if rows.isEmpty { await runtime.refreshDataSourceCollection(windowID: window.windowID, dataSourceRef: dataSourceRef) }
        rows = await runtime.dataSourceCollection(windowID: window.windowID, dataSourceRef: dataSourceRef)
        let stream = await runtime.dataSourceCollectionUpdates(windowID: window.windowID, dataSourceRef: dataSourceRef)
        for await next in stream where !dirty { rows = next }
    }

    private func addRow() {
        let now = Date()
        let formatter = ISO8601DateFormatter()
        var row: [String: JSONValue] = [
            spec.startField ?? "start": .string(formatter.string(from: now)),
            spec.endField ?? "end": .string(formatter.string(from: now.addingTimeInterval(3_600)))
        ]
        if let zone = spec.timeZoneField { row[zone] = .string(TimeZone.current.identifier) }
        rows.append(row)
        dirty = true
    }

    private func dateBinding(index: Int, field: String) -> Binding<Date> {
        Binding(
            get: {
                guard rows.indices.contains(index), case .string(let value)? = rows[index][field] else { return Date() }
                return ISO8601DateFormatter().date(from: value) ?? Date()
            },
            set: { value in
                guard rows.indices.contains(index) else { return }
                rows[index][field] = .string(ISO8601DateFormatter().string(from: value))
                dirty = true
            }
        )
    }

    private func stringBinding(index: Int, field: String, fallback: String) -> Binding<String> {
        Binding(
            get: { rows.indices.contains(index) ? (rows[index][field]?.stringValue ?? fallback) : fallback },
            set: { value in guard rows.indices.contains(index) else { return }; rows[index][field] = .string(value); dirty = true }
        )
    }

    private func rowTimeZone(index: Int) -> TimeZone {
        guard let field = spec.timeZoneField,
              rows.indices.contains(index),
              let identifier = rows[index][field]?.stringValue,
              let zone = TimeZone(identifier: identifier) else { return .current }
        return zone
    }
}

private func scheduleErrorMessage(_ code: String) -> String {
    switch code {
    case "invalid_date": return "Enter a valid start and end time."
    case "invalid_range": return "End time must be after start time."
    case "min_duration": return "The schedule is shorter than the minimum duration."
    case "overlap": return "This schedule overlaps another entry."
    case "timezone": return "Resolve the time-zone or daylight-saving-time value."
    default: return "The schedule is invalid."
    }
}
